jwst/tweakreg/tests/test_tweakreg.py

```diff
@@ -82,11 +94,229 @@ def test_write_catalog(dummy_source_catalog, tmp_cwd):
+    elif meta == "empty":
+        example_input[0].meta.tweakreg_catalog = ""
+
+    # write out the ModelContainer and association (so the association table will be loaded)
+    example_input.save(dir_path=str(custom_catalog_path.parent))
+    asn_data = {
+        'asn_id': 'foo',
+        'asn_pool': 'bar',
+        'products': [
+            {
+                'members': [{'expname': m.meta.filename, 'exptype': 'science'} for m in example_input],
+            },
+        ],
+    }
+
+    if asn == "empty":
+        asn_data['products'][0]['members'][0]['tweakreg_catalog'] = ''
+    elif asn == "valid":
+        asn_data['products'][0]['members'][0]['tweakreg_catalog'] = str(custom_catalog_path.name)
+
+    import json
+    asn_path = custom_catalog_path.parent / 'example_input.json'
+    with open(asn_path, 'w') as f:
+        json.dump(asn_data, f)
+
+    # write out a catfile
+    if catfile != "skip":
+        catfile_path = custom_catalog_path.parent / 'catfile.txt'
+        with open(catfile_path, 'w') as f:
+            if catfile == "valid":
+                f.write(f"{example_input[0].meta.filename} {custom_catalog_path.name}")
+            elif catfile == "empty":
+                f.write(f"{example_input[0].meta.filename}")
+            elif catfile == "invalid":
+                pass
+
+    # figure out how many sources to expect for the model in group 'a' 
+    n_sources = N_EXAMPLE_SOURCES
+    custom_number = N_CUSTOM_SOURCES
+    if not custom:
+        # if use_custom_catalog is False, the custom catalog shouldn't be used
+        n_custom_sources = n_sources
+    else:
+        if catfile == "valid":
+            # for a 'valid' catfile, expect the custom number
+            n_custom_sources = custom_number
+        elif catfile == "invalid":
+            # for an 'invalid' catfile, use_custom_catalog should become disabled
+            n_custom_sources = n_sources
+        elif catfile == "empty":
+            # for a catfile with an 'empty' entry, no custom catalog should be used
+            n_custom_sources = n_sources
+        else:  # catfile == "skip"
+            assert catfile == "skip"  # sanity check
+            # since catfile is not defined, now look at asn_
+            if asn == "valid":
+                # for a 'valid' asn entry, expect the custom number
+                n_custom_sources = custom_number
+            elif asn == "empty":
+                # for a 'empty' asn entry, no custom catalog should be used
+                n_custom_sources = n_sources
+            else:  # asn == "skip"
+                assert asn == "skip"  # sanity check
+                if meta == "valid":
+                    n_custom_sources = custom_number
+                elif meta == "empty":
+                    n_custom_sources = n_sources
+                else:  # meta == "skip"
+                    assert meta == "skip"
+                    n_custom_sources = n_sources
```

Comment:
Unfortunately I agree...

The number of possibilities and combinations is very confusing. Perhaps there are combinations we can remove? Do we know if anyone defines `tweakreg_catalog` in the asn (based on the previous tickets it sounds like `catfile` is used)? Also, do we know if anyone uses the `model.meta.tweakreg_catalog` feature?

I updated the PR to use your suggestion in: https://github.com/spacetelescope/jwst/pull/8424#discussion_r1571309494
See: bcf2ac2f4

Reply:
> The number of possibilities and combinations is very confusing. Perhaps there are combinations we can remove?

Would it be better to split this test into multiple tests that check separately user catalog being provided via asn, meta, or catfile.

> Do we know if anyone defines `tweakreg_catalog` in the asn (based on the previous tickets it sounds like `catfile` is used)?

I don't know. We have no way to collect feedback from the users but I vaguely remember someone asking about it on Slack.

> Also, do we know if anyone uses the `model.meta.tweakreg_catalog` feature?

I don't know this either.

I think we should preserve all these means. However, we should try simplify the logic somehow. It was just hard to look at that huge "if" block.

Reply:
I'm not sure if splitting this into multiple tests would be easier to understand. The main reason being that the options are all inter-related (even for a valid `tweakreg_catalog` in the asn and with `use_custom_catalogs` set to `True` if the user provides in invalid `catfile` no custom catalog will be used). This would mean that if we want to cover all of these combinations each test would have to instrument all options.